### PR TITLE
Fix a use-after-free in MandatoryInlining

### PR DESCRIPTION
**Pull request by meg-gupta**
In recordDeadFunction, we look at operands of an instruction to be deleted,
and add back the defining instruction of the operands to the worklist.
This works in general when we are deleting dead instructions
recursively.
But we also consider, an instruction with only debug uses as dead. So
when we are deleting a debug instruction, we may have already deleted
its operand's defining instruction. So it would be incorrect to add it
to the worklist.

Fixes rdar://72029485

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci test

File: lib/SILOptimizer/Mandatory/MandatoryInlining.cpp [hunk 1]
```diff
@@ -481,6 +481,12 @@ class ClosureCleanup {
       if (!operandVal->getType().is<SILFunctionType>())
         continue;
 
+      // Skip debug instructions
```
Comment (eeckstein) [inline lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:484]: This check should be done at the function entry.

Can you add a bit more information in the comment? E.g. Add the info what you wrote in the PR description.
Also, you should mention that the special handling of debug instruction is done in eraseFromParentWithDebugInsts.

**Review by eeckstein — APPROVED** [review]
This is good to fix the issue.
Though, I have some comments (you can address them in a follow-up commit if you like).

Also, you should add a test.

**Comment by meg-gupta** [comment]
@swift-ci smoke test and merge

**Comment by meg-gupta** [comment]
Thanks @eeckstein . I'll add the test you wrote in a follow on commit.

**Pull request merged**